src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs

```diff
@@ -454,33 +289,93 @@ internal bool IsExpressionBodied
             get
             {
-                if (_lazyType == null)
+                EnsureSignature();
+                return _lazyType.Value;
+            }
+        }
+
+#nullable enable 
+
+        private void EnsureSignature()
+        {
+            if (!_state.HasComplete(CompletionPart.FinishPropertyEnsureSignature))
+            {
+                // If this lock ever encloses a potential call to Debugger.NotifyOfCrossThreadDependency,
+                // then we should call DebuggerUtilities.CallBeforeAcquiringLock() (see method comment for more
+                // details).
+
+                lock (_syntaxRef)
                 {
-                    var diagnostics = DiagnosticBag.GetInstance();
-                    var result = this.ComputeType(binder: null, CSharpSyntaxNode, diagnostics);
-                    if (Interlocked.CompareExchange(ref _lazyType, new TypeWithAnnotations.Boxed(result), null) == null)
+                    if (_state.NotePartComplete(CompletionPart.StartPropertyEnsureSignature))
+                    {
+                        // By setting StartPropertyEnsureSignature, we've committed to doing the work and setting
+                        // FinishPropertyEnsureSignature.  So there is no cancellation supported between one and the other.
+                        var diagnostics = DiagnosticBag.GetInstance();
+                        try
+                        {
+                            EnsureSignatureGuarded(diagnostics);
+                            AddDiagnostics(diagnostics);
+                        }
+                        finally
+                        {
+                            _state.NotePartComplete(CompletionPart.FinishPropertyEnsureSignature);
+                            diagnostics.Free();
+                        }
+                    }
+                    else
                     {
-                        this.AddDeclarationDiagnostics(diagnostics);
+                        // Either (1) this thread is in the process of completing the method,
+                        // or (2) some other thread has beat us to the punch and completed the method.
+                        // We can distinguish the two cases here by checking for the FinishPropertyEnsureSignature
+                        // part to be complete, which would only occur if another thread completed this
+                        // method.
+                        //
+                        // The other case, in which this thread is in the process of completing the method,
+                        // requires that we return here even though the work is not complete. That's because
+                        // signature is processed by first populating the return type and parameters by binding
+                        // the syntax from source.  Those values are visible to the same thread for the purpose
+                        // of computing which methods are implemented and overridden.  But then those values
```

Comment:
Might we be able to use signature-only symbols, rather than partially constructed symbols, to find which methods are implemented and overridden?

Reply:
> Might we be able to use signature-only symbols, rather than partially constructed symbols, to find which methods are implemented and overridden?

I think technically it should be possible to locate overridden/implemented property with an alternative signature representation. A signature-only symbol is one way to do that. Perhaps, if we were implementing this from scratch, we should have taken that route. At this point, however, I prefer sticking to the existing way of doing things, which has been verified to work (in the old constructor and with method symbols). Trying to use signature-only symbols, will likely require changes elsewhere, the symbol itself is getting updated in the process, the symbol itself is getting captured somewhere. For example, in diagnostics, etc. 

---
In reply to: [527209014](https://github.com/dotnet/roslyn/pull/49360#discussion_r527209014) [](ancestors = 527209014)